rust/operator-binary/src/discovery.rs

```diff
@@ -21,12 +21,16 @@ pub fn build_discovery_configmap(
     controller: &str,
     namenode_podrefs: &[HdfsPodRef],
     resolved_product_image: &ResolvedProductImage,
-) -> OperatorResult<ConfigMap> {
+) -> Result<ConfigMap, crate::hdfs_controller::Error> {
     ConfigMapBuilder::new()
         .metadata(
             ObjectMetaBuilder::new()
                 .name_and_namespace(hdfs)
-                .ownerreference_from_resource(hdfs, None, Some(true))?
+                .ownerreference_from_resource(hdfs, None, Some(true))
+                .map_err(|err| Error::ObjectMissingMetadataForOwnerRef {
```

Comment:
@sbernauer I believe we can use .with_context instead of .map_err here, eg:

```
        .with_context(|_| ObjectMissingMetadataForOwnerRefSnafu {
            obj_ref: ObjectRef::from_obj(hdfs),
        })
```

_the source gets set inside with_error, so doesn't need specifying here_

Reply:
Although, it looks like we can simply do:

```
        .context(ObjectMissingMetadataForOwnerRefSnafu {
            obj_ref: ObjectRef::from_obj(hdfs),
        })?
```

So I'm unsure when to use with_context(|e: Error| ...) vs context(...) when you don't need access to the Error

Reply:
Yes, if the field is called `source`, the source error will automatically be picked up. So using `context()` instead of `with_context()` is recommended.

Reply:
Will give that a try next time :+1:

Reply:
done in #454

Reply:
More on this while talking to @nightkr...

Just curious, I see a lot of `with_context` used over the place. Often the error variable is being ignored, so are these equivalent?

```rust
                    .with_context(|| MissingRoleSnafu {
                        role: HdfsRole::NameNode.to_string(),
                    })?;
```

```rust
                    .context(MissingRoleSnafu {
                        role: HdfsRole::NameNode.to_string(),
                    })?;
```

> @nightkr: the difference is that `.context()` is always evaluated, while `.with_context()` is only evaluated if there is an error to add context to
> it doesn't really matter for trivial context selectors, but it lets us avoid `to_string()`/`clone()` calls in the happy case
> but arguably in that case, the error should just take a [HdfsRole in another example] instead of a `String`, resolving that neatly anyway

So just to make sure I understand, if `HdfsRole` implemented `Into<String>`, then we could do this at no expense?

```rust
                    .context(MissingRoleSnafu {
                        role: HdfsRole::NameNode, // into::<String>() will be evaluated if the error actually occurs
                    })?;
```

> @nightkr: yes

Reply:
The difference is that `with_context()` will short-circuit, and only evaluate the ctx in an actual error case, while `context()` will always be evaluated. `with_context()` should be preferred if the ctx is non-trivial (like calling `Clone::clone()` or `Display::to_string()`).